lib/coffee_place/cli.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+require_relative '../coffee_place'
+require_relative './location'
+require_relative './cli_parser'
+require_relative './importer'
+require_relative './search'
+
+module CoffeePlace
+  # Parses and runs CLI application using supplied args.
+  class CLI
+    attr_accessor :cli_opts, :parsed_args
+
+    def initialize(parser: CLIParser.new,
+                   importer: Importer.new)
+      @parser = parser
+      @importer = importer
+    end
+
+    def run(args)
+      cli_opts = parse_cli_options!(args)
+
+      show_help_and_exit! if cli_opts.print_help?
+      show_version_and_exit! if cli_opts.show_version?
+
+      user_location = validate_user_location!(cli_opts)
+      locations = import_locations_from!(cli_opts)
+
+      search = Search.new(locations)
+      closest_locations = search.find_closest_to(user_location, cli_opts.num_results)
+
+      display_closest_locations(closest_locations)
+    end
+
+    private
+
+    def parse_cli_options!(args)
+      result = @parser.parse(args)
+      show_help_and_exit!(exit_status: 1) unless result.success?
+
+      cli_opts = result.value
+
+      # We must have [lat, lon, filename] as args
+      unless cli_opts.num_args == 3
+        puts 'Incorrect number of CLI arguments passed in'
+        show_help_and_exit!(exit_status: 1)
+      end
+
+      cli_opts
+    end
+
+    def validate_user_location!(cli_opts)
+      lat, lon = cli_opts.remaining
+      location_result = Location.validate(lat: lat, lon: lon, name: 'Current user')
```

Comment:
What is the reasoning behind the validate method to return, in case of success, an instance of `CoffeePlace::Location`? From the naming, I'd not expect it.

Reply:
You're right, this method does more than what is expected.
I'll try to refactor this part and push another commit.

Reply:
Update: fixed by [482136f9](https://github.com/Agilefreaks/test_oop/pull/23/commits/482136f90b3e7cd0c8681e32000b92e1ab97b64c)